### PR TITLE
Linux broker: part 1

### PR DESCRIPTION
**Pull request by karthiknadig**
Implements interpreter detection.
Implements system information.

### CODE REVIEW

File: src/Linux/Core/Impl/IO/FileSystem.cs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+namespace Microsoft.R.Common.Core.Linux {
+    public class FileSystem : IFileSystem {
+
+        public IFileSystemWatcher CreateFileSystemWatcher(string path, string filter) => new FileSystemWatcherProxy(path, filter);
+
+        public IDirectoryInfo GetDirectoryInfo(string directoryPath) => new DirectoryInfoProxy(directoryPath);
+
+        public bool FileExists(string path) => File.Exists(path);
+
```
Comment (MikhailArkhipov) [inline src/Linux/Core/Impl/IO/FileSystem.cs]: Looks like mostly same thing as Windows. Perhaps just have common in Core or at least a base class if implementations are slightly different?

Comment (MikhailArkhipov) [inline src/Linux/Core/Impl/IO/FileSystem.cs]: Also, is Linux Core would be the same as Mac? Maybe name it `.Unix` then?

File: src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            string cpuLoadInfo = "/proc/stat";
+            var lines = _fs.FileReadAllLines(cpuLoadInfo).ToArray();
+            var split = lines[0].Split(_cpuInfoSplitter, StringSplitOptions.RemoveEmptyEntries);
+            long user = 0, nice = 0, system = 0, idle = 0, iowait = 0, irq = 0, softirq = 0, steal = 0;
+            // split[0] == "cpu" so ignore the first item;
+            if (split.Length > 1 && !long.TryParse(split[1], out user)) {
+                user = 0;
+            }
+
+            if (split.Length > 2 && !long.TryParse(split[2], out nice)) {
```
Comment (MikhailArkhipov) [inline src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs]: `out long nice` (C# 7)

File: src/Linux/Core/Impl/OS/InstalledPackageInfo.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            if (fs.FileExists(path)) {
+                return fs.FileReadAllLines(path);
+            }
+
+            path = $"/var/lib/dpkg/info/{PackageName}:{Architecture}.list";
+            if (fs.FileExists(path)) {
+                return fs.FileReadAllLines(path);
+            }
+
+            var files = Directory.GetFiles("/var/lib/dpkg/info", $"{PackageName}*.list", SearchOption.TopDirectoryOnly);
```
Comment (MikhailArkhipov) [inline src/Linux/Core/Impl/OS/InstalledPackageInfo.cs]: `fs.GetFiles`? Add it if it is not there

File: src/Linux/Core/Impl/OS/InstalledPackageInfo.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            if (files.Length > 0) {
+                return fs.FileReadAllLines(files[0]);
+            }
+
+            return new string[0];
+        }
+
+        public static IEnumerable<InstalledPackageInfo> GetPackages(IFileSystem fs) {
```
Comment (MikhailArkhipov) [inline src/Linux/Core/Impl/OS/InstalledPackageInfo.cs]: Needs some unit test

File: src/Linux/Core/Impl/OS/InstalledPackageInfo.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                }
+            } catch (Exception ex) when (!ex.IsCriticalException()) {
+            }
+            return list;
+        }
+
+        private static string GetPart(string[] lines, string part, ref int index) {
+            if (index < lines.Length) {
+                string line = lines[index];
+                while (!line.StartsWith(part)) {
```
Comment (MikhailArkhipov) [inline src/Linux/Core/Impl/OS/InstalledPackageInfo.cs]: `StartsWithOrdinal`

File: src/Linux/Core/Impl/OS/InstalledPackageInfo.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                        return null;
+                    }
+                    line = lines[index];
+                }
+                return line.Substring(part.Length).Trim();
+            }
+            return null;
+        }
+
+        static string[] versionSeperators = { ".", "-", "~", " ", ":", "+" };
```
Comment (MikhailArkhipov) [inline src/Linux/Core/Impl/OS/InstalledPackageInfo.cs]: `separators`

File: src/Linux/Core/Impl/OS/InstalledPackageInfo.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        private static Version ParseVersion(string version) {
+            // this is linux package version string which might contain text
+            // e.g, 3.2.3~pre-6ubuntu8 OR 8c-0ubuntu1 OR 2016asdc-ubuntu02
+            // extract major and minor version, only if it exists
+            string[] split = version.Split(versionSeperators, StringSplitOptions.RemoveEmptyEntries);
+
+            Version v;
+            if (split.Length >= 4 && System.Version.TryParse($"{split[0]}.{split[1]}.{split[2]}.{split[3]}", out v)) {
+                return v;
```
Comment (MikhailArkhipov) [inline src/Linux/Core/Impl/OS/InstalledPackageInfo.cs]: `out Version v`

File: src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            long initIdleTime = initial.idle + initial.iowait;
+            long currentIdleTime = current.idle + current.iowait;
+            long deltaIdleTime = currentIdleTime - initIdleTime;
+
+            long deltaUsage = deltaTimeSinceBoot - deltaIdleTime;
+            return deltaUsage / deltaTimeSinceBoot;
+        }
+
+        public (long TotalVirtualMemory, long FreeVirtualMemory, long TotalPhysicalMemory, long FreePhysicalMemory) GetMemoryInformation() {
+            var meminfo = GetMemInfo();
```
Comment (MikhailArkhipov) [inline src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs]: May need to handle `IOException` and `UnauthorizedAccessException` if `ReadAllLines` throws

File: src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+        public (string VideoCardName, long VideoRAM, string VideoProcessor) GetVideoControllerInformation() {
+            return ("Unknown", 0, "Unknown");
+        }
+
+        private static char[] _cpuInfoSplitter = new char[] { ' ', ':', '\t' };
+        private (long user, long nice, long system, long idle, long iowait, long irq, long softirq, long steal) GetCurrentCpuUsage() {
+            string cpuLoadInfo = "/proc/stat";
+            var lines = _fs.FileReadAllLines(cpuLoadInfo).ToArray();
+            var split = lines[0].Split(_cpuInfoSplitter, StringSplitOptions.RemoveEmptyEntries);
```
Comment (MikhailArkhipov) [inline src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs]: Will crash if array is empty

File: src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+            if (split.Length > 6 && !long.TryParse(split[6], out irq)) {
+                irq = 0;
+            }
+
+            if (split.Length > 7 && !long.TryParse(split[7], out softirq)) {
+                softirq = 0;
+            }
+
+            if (split.Length > 8 && !long.TryParse(split[8], out steal)) {
+                steal = 0;
```
Comment (MikhailArkhipov) [inline src/Linux/Host/Broker/Impl/Services/LinuxSystemInfoService.cs]: I believe if `TryParse` fails, result will be 0 anyway

File: src/Linux/Host/Broker/Impl/Startup/LinuxStartup.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public sealed class LinuxStartup : Startup {
+        public LinuxStartup(ILoggerFactory loggerFactory, IConfigurationRoot configuration) : base(loggerFactory, configuration) { }
+
+        public override void ConfigureServices(IServiceCollection services) {
+            base.ConfigureServices(services);
+
+            var fs = new FileSystem();
+            services.AddSingleton<IFileSystem>(fs)
+                .AddSingleton<IAuthenticationService, LinuxAuthenticationService>()
```
Comment (MikhailArkhipov) [inline src/Linux/Host/Broker/Impl/Startup/LinuxStartup.cs]: If it is expected to be frequently used service, you can create and add actual instance and avoid reflection-based creation. I.e. .AddService(new Foo())

File: src/Linux/R/Interpreters/Impl/RInstallation.cs [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+using Microsoft.R.Common.Core.Linux;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.R.Interpreters
+{
+    public sealed class RInstallation : IRInstallationService
+    {
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Impl/RInstallation.cs]: Formatting

File: src/Linux/R/Interpreters/Impl/RInstallation.cs [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+
+        public IRInterpreterInfo CreateInfo(string name, string path)
+        {
+            throw new NotImplementedException();
+        }
+
+        public IEnumerable<IRInterpreterInfo> GetCompatibleEngines(ISupportedRVersionRange svl = null)
+        {
+            var packagesInfo = InstalledPackageInfo.GetPackages(_fileSystem);
+            List<IRInterpreterInfo> interpreters = new List<IRInterpreterInfo>();
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Impl/RInstallation.cs]: `var`

File: src/Linux/R/Interpreters/Impl/RInstallation.cs [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+                list.Add(new RInterpreterInfo($"Microsoft R Open '{package.Version}'", Path.GetDirectoryName(libRPath), package.Version, package.GetVersion(), _fileSystem));
+            }
+
+            return list;
+        }
+
+        private IEnumerable<IRInterpreterInfo> GetInstalledCranR(IEnumerable<InstalledPackageInfo> packagesInfo, ISupportedRVersionRange svl)
+        {
+            var list = new List<IRInterpreterInfo>();
+            var selectedPackages = packagesInfo.Where(p => p.PackageName.StartsWithIgnoreCase("microsoft-r-open-mro") && svl.IsCompatibleVersion(p.GetVersion()));
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Impl/RInstallation.cs]: CRAN probably does not start with `microsoft` ;-)

File: src/Linux/R/Interpreters/Impl/RInstallation.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            {
+                var files = package.GetPackageFiles(_fileSystem);
+                string libRPath = GetLibRPath(files, _fileSystem);
+                list.Add(new RInterpreterInfo($"Microsoft R Open '{package.Version}'", Path.GetDirectoryName(libRPath), package.Version, package.GetVersion(), _fileSystem));
+            }
+
+            return list;
+        }
+
+        private IEnumerable<IRInterpreterInfo> GetInstalledCranR(IEnumerable<InstalledPackageInfo> packagesInfo, ISupportedRVersionRange svl)
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Impl/RInstallation.cs]: Looks pretty close to the above, maybe make some common method and pass prefix?

File: src/Linux/R/Interpreters/Impl/RInstallation.cs [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+namespace Microsoft.R.Interpreters
+{
+    public sealed class RInstallation : IRInstallationService
+    {
+
+        private const string _mroDirectory = "/usr/lib64/microsoft-r";
+        private const string _mroName = "Microsoft R Open";
+        private readonly IFileSystem _fileSystem;
+
+        public RInstallation() :
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Impl/RInstallation.cs]: Can use unit tests too

File: src/Linux/R/Interpreters/Impl/RInterpreterInfo.cs [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+                            string.Format(CultureInfo.InvariantCulture, Resources.Error_UnsupportedRVersion,
+                            Version.Major, Version.Minor, Version.Build, svr.MinMajorVersion, svr.MinMinorVersion, "*",
+                            svr.MaxMajorVersion, svr.MaxMinorVersion, "*"), MessageButtons.OK);
+                    }
+                } else {
+                    ui?.ShowMessage(string.Format(CultureInfo.InvariantCulture, Resources.Error_CannotFindRBinariesFormat, InstallPath), MessageButtons.OK);
+                }
+            } catch (Exception ex) when (ex is IOException || ex is ArgumentException || ex is UnauthorizedAccessException) {
+                ui?.ShowErrorMessage(
+                    string.Format(CultureInfo.InvariantCulture, Resources.Error_ExceptionAccessingPath, InstallPath, ex.Message));
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Impl/RInterpreterInfo.cs]: `Resources.Error_ExceptionAccessingPath.FormatInvariant()`

File: src/Linux/R/Interpreters/Impl/RInterpreterInfo.cs [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+            }
+
+            return _isValid.Value;
+        }
+
+        private string GetRLibPath()
+        {
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Impl/RInterpreterInfo.cs]: Formatting

File: src/Linux/Core/Impl/OS/InstalledPackageInfo.cs (hunk 4 above)
Comment (karthiknadig) [inline src/Linux/Core/Impl/OS/InstalledPackageInfo.cs]: Added.

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+using System;
+using System.Linq;
+using FluentAssertions;
+using Microsoft.Common.Core;
+using Microsoft.UnitTests.Core.Linux;
+using Xunit;
+
+namespace Microsoft.R.Interpreters.Linux.Test {
+    public class RInstallationTest {
+        [Fact]
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]: I think we use `[Test]` but I guess that's OK

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+        public void RInstallationBasicTest() {
+            var fs = new TestFileSystem();
+            var rInstallation = new RInstallation(fs);
+            var installs = rInstallation.GetCompatibleEngines();
+            installs.Count().Should().Be(2);
+
+            // test MRO
+            var mro = installs.Where(i => i.Name.StartsWithOrdinal("Microsoft"));
+            mro.Count().Should().Be(1);
```
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]: `.Should().HaveCount(1)`

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs (hunk 19 above)
Comment (karthiknadig) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]: `[Fact]` was the default for .Net Core test project.

**Comment by MikhailArkhipov** [comment]
👍 @AlexanderSher may want to look at the test code

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs (hunk 19 above)
Comment (MikhailArkhipov) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]: OK

File: src/Windows/R/Components/Impl/History/Implementation/RHistoryProvider.cs [hunk 21]
```diff
@@ -53,7 +53,7 @@ internal class RHistoryProvider : IRHistoryProvider {
         public IRHistory CreateRHistory(IRInteractiveWorkflowVisual interactiveWorkflow) {
             var contentType = _contentTypeRegistryService.GetContentType(RHistoryContentTypeDefinition.ContentType);
             var textBuffer = _textBufferFactory.CreateTextBuffer(contentType);
-            var history = new RHistory(interactiveWorkflow, textBuffer, new FileSystem(), _settings, _editorOperationsFactory, _rtfBuilderService, () => RemoveRHistory(textBuffer));
+            var history = new RHistory(interactiveWorkflow, textBuffer, new WindowsFileSystem(), _settings, _editorOperationsFactory, _rtfBuilderService, () => RemoveRHistory(textBuffer));
```
Comment (AlexanderSher) [inline src/Windows/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: It's better to get it from `ICoreShell` in constructor

Comment (AlexanderSher) [inline src/Windows/R/Components/Impl/History/Implementation/RHistoryProvider.cs]: `coreShell.FileSystem()`

File: src/Windows/R/Components/Impl/InteractiveWorkflow/Commands/SourceRScriptCommand.cs [hunk 22]
```diff
@@ -17,13 +17,13 @@ public sealed class SourceRScriptCommand : IAsyncCommand {
...
         private readonly bool _echo;
-        private readonly FileSystem _fs;
+        private readonly WindowsFileSystem _fs;
 
         public SourceRScriptCommand(IRInteractiveWorkflowVisual interactiveWorkflow, IActiveWpfTextViewTracker activeTextViewTracker, bool echo) {
             _interactiveWorkflow = interactiveWorkflow;
             _activeTextViewTracker = activeTextViewTracker;
             _echo = echo;
-            _fs = new FileSystem();
+            _fs = new WindowsFileSystem();
```
Comment (AlexanderSher) [inline src/Windows/R/Components/Impl/InteractiveWorkflow/Commands/SourceRScriptCommand.cs]: Same thing here. `ICoreShell` can be accessed from `_interactiveWorkflow`

File: src/Windows/R/Components/Impl/InteractiveWorkflow/Commands/SourceRScriptCommand.cs [hunk 23]
```diff
@@ -67,7 +67,7 @@ public sealed class SourceRScriptCommand : IAsyncCommand {
 
             var session = _interactiveWorkflow.RSession;
             if (session.IsRemote) {
-                using (DataTransferSession dts = new DataTransferSession(_interactiveWorkflow.RSession, new FileSystem())) {
+                using (DataTransferSession dts = new DataTransferSession(_interactiveWorkflow.RSession, new WindowsFileSystem())) {
```
Comment (AlexanderSher) [inline src/Windows/R/Components/Impl/InteractiveWorkflow/Commands/SourceRScriptCommand.cs]: `_fs`

File: src/Windows/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs [hunk 24]
```diff
@@ -84,7 +84,7 @@ public sealed class RInteractiveEvaluator : IInteractiveEvaluator {
             try {
                 if (!Session.IsHostRunning) {
                     var startupInfo = new RHostStartupInfo(_settings.CranMirror, _settings.WorkingDirectory, _settings.RCodePage, _terminalWidth, !isResetting, true, true);
-                    await Session.EnsureHostStartedAsync(startupInfo, new RSessionCallback(CurrentWindow, Session, _settings, _coreShell, new FileSystem()));
+                    await Session.EnsureHostStartedAsync(startupInfo, new RSessionCallback(CurrentWindow, Session, _settings, _coreShell, new WindowsFileSystem()));
```
Comment (AlexanderSher) [inline src/Windows/R/Components/Impl/InteractiveWorkflow/Implementation/RInteractiveEvaluator.cs]: `_coreShell.FileSystem()`
And in other places

File: src/Package/Impl/Sql/Publish/SqlPublshOptionsDialog.xaml.cs [hunk 25]
```diff
@@ -35,7 +35,7 @@ public partial class SqlPublshOptionsDialog : PlatformDialogWindow {
 
         public static async Task<SqlPublshOptionsDialog> CreateAsync(
             ICoreShell shell, IProjectSystemServices pss, IProjectConfigurationSettingsProvider pcsp, ISettingsStorage settings) {
-            var dialog = await CreateAsync(shell, pss, new FileSystem(), pcsp, settings);
+            var dialog = await CreateAsync(shell, pss, new WindowsFileSystem(), pcsp, settings);
```
Comment (AlexanderSher) [inline src/Package/Impl/Sql/Publish/SqlPublshOptionsDialog.xaml.cs]: `fs` parameter isn't used in `SqlPublshOptionsDialog` - you can reduce the whole thing to one constructor

File: src/Package/Impl/Sql/Commands/PublishSProcCommand.cs [hunk 26]
```diff
@@ -28,7 +28,7 @@ internal sealed class PublishSProcCommand : ICommandGroupHandler {
 
         [ImportingConstructor]
         public PublishSProcCommand(ICoreShell shell, IProjectSystemServices pss, IDacPackageServicesProvider dacServicesProvider, ISettingsStorage settings) :
-            this(shell, pss, new FileSystem(), dacServicesProvider, settings) {
+            this(shell, pss, new WindowsFileSystem(), dacServicesProvider, settings) {
```
Comment (AlexanderSher) [inline src/Package/Impl/Sql/Commands/PublishSProcCommand.cs]: `shell.FileSystem()`

File: src/Linux/Core/Impl/IO/UnixFileSystem.cs [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.IO;
+using Microsoft.Extensions.FileSystemGlobbing;
+
+namespace Microsoft.R.Common.Core.Linux {
```
Comment (AlexanderSher) [inline src/Linux/Core/Impl/IO/UnixFileSystem.cs]: I believe we have decided to use one common namespace for both xplat types and platform-specific types. So it should be `Microsoft.Common.Core.IO`.

File: src/Common/Core/Impl/IO/FileSystem.cs [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Compression;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.IO;
+using Microsoft.Extensions.FileSystemGlobbing;
+
+namespace Microsoft.Common.Core {
```
Comment (AlexanderSher) [inline src/Common/Core/Impl/IO/FileSystem.cs]: `Microsoft.Common.Core.IO`

File: src/Linux/Core/Impl/Microsoft.R.Common.Core.Linux.csproj [hunk 29]
```diff
@@ -0,0 +1,14 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard1.6</TargetFramework>
+    <AssemblyName>Microsoft.R.Common.Core.Linux</AssemblyName>
```
Comment (AlexanderSher) [inline src/Linux/Core/Impl/Microsoft.R.Common.Core.Linux.csproj]: You can set default namespace here:
`<RootNamespace>Microsoft.Common.Core</RootNamespace>`

File: src/Linux/Core/Impl/OS/InstalledPackageInfo.cs [hunk 30]
```diff
@@ -0,0 +1,112 @@
...
+
+using Microsoft.Common.Core;
+using Microsoft.Common.Core.IO;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.R.Common.Core.Linux {
```
Comment (AlexanderSher) [inline src/Linux/Core/Impl/OS/InstalledPackageInfo.cs]: `Microsoft.Common.Core.OS`

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs (hunk 19 above)
Comment (AlexanderSher) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]: > `[Fact]` was the default for .Net Core test project.

Yes it is, though we use our own attributes. They are defined in .net framework project, so you won't be able to use them. I've created a task #3489

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs (hunk 20 above)
Comment (AlexanderSher) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]:
```cs
installs.Should.ContainSingle(i => i.Name.StartsWithOrdinal("Microsoft"))
    .Which().Version.Should().Be(new Version(3, 3, 3))
```

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs [hunk 31]
```diff
@@ -0,0 +1,41 @@
...
+            var mroInterpreter = mro.First();
+            mroInterpreter.Version.Should().Be(new Version(3, 3, 3));
+
+            // test CRAN R
+            var cranR = installs.Where(i => i.Name.StartsWithOrdinal("R "));
```
Comment (AlexanderSher) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]:
```cs
installs.Should.ContainSingle(i => i.Name.StartsWithOrdinal("R"))
    .Which().Version.Should().Be(new Version(3, 2, 3, 4))
```

File: src/Linux/R/Interpreters/Test/RInstallationTest.cs [hunk 32]
```diff
@@ -0,0 +1,41 @@
...
+using Microsoft.UnitTests.Core.Linux;
+using Xunit;
+
+namespace Microsoft.R.Interpreters.Linux.Test {
+    public class RInstallationTest {
+        [Fact]
+        public void RInstallationBasicTest() {
+            var fs = new TestFileSystem();
+            var rInstallation = new RInstallation(fs);
+            var installs = rInstallation.GetCompatibleEngines();
```
Comment (AlexanderSher) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]: Also, if the order is determinate, the whole test can be reduced to:
```cs
new RInstallation().GetCompatibleEngines(svr).Should().Equal(
    new[] {
        new {Name = "Microsoft", Version = new Version(3, 3, 3)},
        new {Name = "R", Version = new Version(3, 3, 3)}
    },
    (interpreter, expected) => interpreter.Name.StartsWith(expected.Name) && interpreter.Version == expected.Version);
```

Comment (karthiknadig) [inline src/Linux/R/Interpreters/Test/RInstallationTest.cs]: The current test file results in this order, if we update the test file in the future the order may change. In the Linux package status file, the order is not determinate.

**Comment by karthiknadig** [comment]
I have made the changes recommended by @AlexanderSher.

**Pull request merged**